### PR TITLE
python3Packages.gluonts: init at 0.13.5

### PR DESCRIPTION
**Pull request by mbalatsko**
## Description of changes

Introducing `gluonts` python package, also adding its dependency `cpflows` package

- `[R, mxnet]` extensions are for now omitted, added comments in the package definition.
- Original [cpflows package](https://github.com/CW-Huang/CP-Flow) is not a real python package. `gluonts` team created a special [fork](https://github.com/lostella/CP-Flow/tree/gluonts-dependency), which turns this repo into a python package. This fork was used to introduce `cpflows` python package in nixpkgs.

Adding one of the requirements needed for https://github.com/NixOS/nixpkgs/issues/247089 . I want to separate mindsdb requirements initialization into separate PRs bacause adding it to nixpkgs requires more than 15 new packages. I believe this way it would be easier to review.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mbalatsko** [comment]
Results of `nixpkgs-review pr 255682`: 
```
3 packages built:
python310Packages.cpflows python310Packages.gluonts python311Packages.cpflows
```

**Comment by mbalatsko** [comment]
@GrahamcOfBorg build python310Packages.cpflows python310Packages.gluonts python310Packages.cpflows python311Packages.cpflows

File: pkgs/development/python-modules/cpflows/default.nix [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    owner = "lostella";
+    repo = "CP-Flow";
+    rev = "96307119fe734c3226b0f685ddaf10f0004fb177"; # gluonts-dependency branch head
+    hash = "sha256-VtDDjpOv85GbBqv0wyyw4A4XOV+8aoLz8XGM0IB2Fec=";
+  };
+
+  patches = [
+    ./fix-dependencies.patch
+  ];
+
```
Comment (fabaff) [inline pkgs/development/python-modules/cpflows/default.nix:44]:
```suggestion
  patches = [
    ./fix-dependencies.patch
  ];
```

Could you please check if it's possible to use `pythonRelaxDepsHook` here? Editing/creating patch files for updates is always a bit of a hassle from my point of view. If the hook doesn't work a `postPatch` section with `substituteInPlace` could also do the job.

File: pkgs/development/python-modules/gluonts/default.nix [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    prophet
+    s3fs
+    fsspec
+    cpflows
+    numba
+    statsmodels
+    plotly
+    matplotlib
+    holidays
+  ];
```
Comment (fabaff) [inline pkgs/development/python-modules/gluonts/default.nix:75]: Some of the requirement (https://github.com/awslabs/gluonts/blob/dev/setup.py#L122C1-L131C32) could go into `passthru.optional-dependencies`.

File: pkgs/development/python-modules/cpflows/default.nix (hunk 1 above)
Comment (mbalatsko) [reply pkgs/development/python-modules/cpflows/default.nix]: Fixed

File: pkgs/development/python-modules/gluonts/default.nix (hunk 2 above)
Comment (mbalatsko) [reply pkgs/development/python-modules/gluonts/default.nix]: Fixed

**Comment by mbalatsko** [comment]
I also used newer 0.13.5 version of `gluonts`, which was released in a meanwhile

**Comment by mbalatsko** [comment]
Result of `nixpkgs-review pr 255682` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>python310Packages.cpflows</li>
    <li>python310Packages.cpflows.dist</li>
    <li>python310Packages.gluonts</li>
    <li>python310Packages.gluonts.dist</li>
    <li>python311Packages.cpflows</li>
    <li>python311Packages.cpflows.dist</li>
  </ul>
</details>

**Comment by happysalada** [comment]
This looks good, i only have a single comment.

**Comment by mbalatsko** [comment]
> This looks good, i only have a single comment.

You have probably forgotten to post it ;)

**Pull request closed without merging**